File: Javascript/stack-max.js
// Define stack constructor function, with maxStack and max function inside of it

function Stack() {
  this.stack = [];
  this.maxStack = []

  this.push = function(data) {
    this.stack.push(data);
    if (this.maxStack.length === 0 || this.maxStack[this.maxStack.length - 1] < data) {
      this.maxStack.push(data);
    } else {
      this.maxStack.push(this.maxStack[this.maxStack.length - 1]);
    }
  }

  this.pop = function() {
    this.maxStack.pop();
    return this.stack.pop();
  }

  this.peek = function() {
    return this.stack[this.stack.length - 1];
  }

  this.max = function() {
    return this.maxStack[this.stack.length - 1];
  }
}

// Create a stack

var stack = new Stack()

stack.push(5);
stack.push(7);
stack.push(6);
stack.push(8);
stack.push(6);

// Test max function at different points

console.log(stack.max());
stack.pop();
stack.pop();
console.log(stack.max());
stack.push(10);
console.log(stack.max());
stack.push(7);
console.log(stack.max());

File: Python/stack-max.py
# This method is used to find the highest value contained
# within a stack.

class Stack(object):
    def __init__(self):
        self.stack = []
        self.max_stack = []

    def push(self, data):
        self.stack.append(data)
        if len(self.max_stack) == 0 or data > self.max_stack[len(self.max_stack) - 1]:
            self.max_stack.append(data)
        else:
            self.max_stack.append(self.max_stack[len(self.max_stack) - 1])

    def pop(self):
        self.stack.pop()
        self.max_stack.pop()

    def peek(self):
        return self.stack[-1]

    def max(self):
        return self.max_stack[-1]

# Generate a stack

stack = Stack()
stack.push(5)
stack.push(7)
stack.push(4)
stack.push(8)
stack.push(5)
stack.push(6)

# Test max function at different points

print stack.max()
stack.pop()
print stack.max()
stack.pop()
stack.pop()
print stack.max()
stack.pop()
print stack.max()
stack.pop()
print stack.max()
